lib/SILOptimizer/Utils/ConstExpr.cpp

```diff
@@ -523,6 +577,152 @@ SymbolicValue ConstExprFunctionState::getConstantValue(SILValue value) {
   return result;
 }
 
+/// Given an aggregate value like {{1, 2}, 3} and an access path like [0,1], and
+/// a scalar like 4, return the aggregate value with the indexed element
+/// replaced with its specified scalar, producing {{1, 4}, 3} in this case.
+/// If `writeOnlyOnce` is true, and the target aggregate element to update
+/// already has a constant value, fail on the update.
+///
+/// This returns true on failure and false on success.
+///
+static bool updateIndexedElement(SymbolicValue &aggregate,
+                                 ArrayRef<unsigned> indices,
+                                 SymbolicValue scalar, Type type,
+                                 bool writeOnlyOnce,
+                                 ASTContext &astContext) {
+  // We're done if we've run out of indices.
+  if (indices.empty()) {
+    aggregate = scalar;
+    return false;
+  }
+
+  // If we have an uninit memory, then scalarize it into an aggregate to
+  // continue.  This happens when memory objects are initialized piecewise.
```

Comment:
Is it possible to write test that fails when this scalarization is removed?

Reply:
Yep, I added a SIL test for it!